### PR TITLE
fix: reset sync_head to header_head on initial transition to HeaderSync (#1524)

### PR DESCRIPTION
**Pull request by garyyu**
The root cause analysis is in https://github.com/mimblewimble/grin/issues/1524.

The fix solution:

1. In `sync.rs`, allow `header_sync()` always execute regardless the state of `SyncStatus`, even on `NoSync` state, but in 60s periodic instead of 10s.
2. The key part. In `adapter.rs`, `headers_received()` function, in case of "All known, ignoring", check the sync head and update it if needed.
```
	let tip = w(&self.chain).get_sync_head().unwrap();
	if bhs_last.height > tip.height || w(&self.chain).get_block_header(&tip.hash()).is_err()
	{
		let res = w(&self.chain).update_sync_head(&bhs_last, self.chain_opts());
	}
``` 

**Updated on 18th Sep. :**
The final fix solution in this PR is far from above, please look the comment here:  https://github.com/mimblewimble/grin/pull/1531#issuecomment-422221501

### CODE REVIEW

**Comment by garyyu** [comment]
And it include another fix, with 3 new lines of code in `sync.rs`:
```
				match status {								match status {
					SyncStatus::TxHashsetDownload => (),									SyncStatus::TxHashsetDownload => (),
+					SyncStatus::HeaderSync { .. }
+					| SyncStatus::NoSync
+					| SyncStatus::Initial => {
```

This fix can avoid the sync state switching between `BodySync` and `HeaderSync`. 
In the old code, there's a risk on `pipe.rs` `update_head()` function,  it's possible sometimes save_head and sometimes not saving head, then the database header chain could be not continuous.
```
		if ctx.opts.contains(Options::SYNC) { 
			batch
				.save_body_head(&tip)
				.map_err(|e| ErrorKind::StoreErr(e, "pipe save body".to_owned()))?;
		} else {          
			batch
				.save_head(&tip)
				.map_err(|e| ErrorKind::StoreErr(e, "pipe save head".to_owned()))?;
		}
```

**Comment by antiochp** [comment]
> in case of "All known, ignoring", check the sync head and update it if needed.

Maybe I'm misunderstanding but this doesn't make intuitive sense to me.

If all the headers are known and we are ignoring them, there should be _nothing_ to do.

**Comment by garyyu** [comment]
Thanks @antiochp  for the review.
"All known, ignoring" only means all received headers are already in header chain, but they could come from block broadcasting, not from previous `HeaderSync`, and in this case, only the `sync head` is not updated and left behind. That's the problem what we're fixing in this PR.

The detail analysis is in the #1514.

**Comment by antiochp** [comment]
I still assert that it would be unexpected and surprising to find a "All known, ignoring" during a header sync produced a side-effect that updated the `sync_head` to a header unrelated to the headers we just ignored.

The original behavior here (at some point) was to update the `sync_head` when we transitioned from `NoSync` to `HeaderSync`. At least I'm pretty sure we did that at some point.

The `sync_head` should reflect the last header that was involved in sync. If we received a header as part of a block broadcast then we should not be updating the `sync_head`.
And we should only need to update the `sync_head` if we later discover we are out of sync - at which point we start a new header sync.
That was the original intent behind the introduction of the `sync_head` tag.

Remember there can be several long-running forks.
The chain proper may be on one fork (the most work chain known so far).
The header chain may have diverged from this - a peer may have advertised a fork with more work, which we are working toward (we do not yet have full blocks for this).
And additionally the `sync_head` may deviate from one or both of these as we may not have received all the headers during sync on this fork.

See here for some of the early design around the `sync_head` and how it related to the `head` and the "header head" and the "chain head" - 
https://github.com/mimblewimble/grin/blob/master/doc/chain/blocks_and_headers.md

The reason we want to avoid updating the `sync_head` too often or too aggressively is we risk flip-flopping between multiple chain forks if we are receiving headers from peers on different forks.
It should only be updated at the start of a header sync and as the sync progresses.
And the target it is aiming for (the "header head") should only ever update when a peer advertises a fork with more work and we have verified all the headers up to and including the one advertised.

**Comment by antiochp** [comment]
I'm not sure I follow this either - 
> In the old code, there's a risk on pipe.rs update_head() function, it's possible sometimes save_head and sometimes not saving head, then the database header chain could be not continuous.

How would the header chain not be continuous?
We will not necessarily call `save_head` on every intermediate header in a particular fork, only on the the one that makes that fork become the most work (header) chain.

**Comment by garyyu** [comment]
Perhaps let's finish the main point firstly, and come to that minor fix (for the risk part) later.

I understand you almost know every detail of those code @antiochp  👍  😄 
but I still want to suggest you read the changed code firstly, I feel you didn't look it:)

Please let me give answers (or question) under the quote:

> I still assert that it would be unexpected and surprising to find a "All known, ignoring" during a header sync produced a side-effect that updated the sync_head to a header unrelated to the headers we just ignored.

No, I don't think so. "All known, ignoring" has a key point at **All known**, that means even we do not ignore them and check&save all these headers again, the result will be same as ignoring them. The only possible missing processing is: ignore and return here will cause `sync head` not updated according to these latest received Headers from HeaderSync!  that's what we're fixing in this PR.

To explain why `the result will be same as ignoring them`, please look these 3 lines: https://github.com/mimblewimble/grin/blob/21767bceec46049ad5e8c9344770187f905c31c7/servers/src/common/adapters.rs#L263-L265
This checking is to check whether the last header in received Headers is already in header chain: `w(&self.chain).get_block_header(&bhs_last.hash())`.

> The original behavior here (at some point) was to update the sync_head when we transitioned from NoSync to HeaderSync. At least I'm pretty sure we did that at some point.

Yes, exactly. And now also, we still keep this original behavior, no change.

> The sync_head should reflect the last header that was involved in sync. If we received a header as part of a block broadcast then we should not be updating the sync_head.

Yes, exactly. And now it's same, no change.

> And we should only need to update the sync_head if we later discover we are out of sync - at which point we start a new header sync.

Yes, and no change.

> The reason we want to avoid updating the sync_head too often or too aggressively is we risk flip-flopping between multiple chain forks if we are receiving headers from peers on different forks.

Hmm, good point! this is what I missed. So, does that mean I'd better remove the new added `Independent HeaderSync` (in every 60s)?

> It should only be updated at the start of a header sync and as the sync progresses.

Not sure I can understand. Why `only at the start of a header sync`?

> And the target it is aiming for (the "header head") should only ever update when a peer advertises a fork with more work and we have verified all the headers up to and including the one advertised.

Not sure  I can understand.  Why you switch to talk about `header head`? It look like irrelevant to this PR, does it?

**Comment by antiochp** [comment]
> This checking is to check whether the last header in received Headers is already in header chain: w(&self.chain).get_block_header(&bhs_last.hash()).

I think this may be part of the misunderstanding. 
This checks if we have seen and processed these headers before, _not_ that they are on the header chain. All of these headers could be on a fork and not on the chain itself.
We could be receiving batches of 500 headers during a header sync that are known to not exist on the current chain, but that are on a fork that _may_ become the header chain.
We may also be receiving full blocks via gossip/broadcast that are not on the main chain, but we still want to accept and process them (within reason) as that fork may eventually become the main chain with most work.

> Hmm, good point! this is what I missed. So, does that mean I'd better remove the new added Independent HeaderSync (in every 60s)?

I think so yes. We don't want a sync to be running periodically in the background all the time.
We only want to initiate a sync if we know for sure that a peer is advertising at least several blocks ahead of where we believe the head to be.

> Not sure I can understand. Why only at the start of a header sync?

We only want to reset the `sync_header` when we start a new sync. If we reset it outside of sync then we risk flip-flopping between two or more different forks.

**Comment by garyyu** [comment]
@ignopeverell: 
> so does that mean the sync_head wasn't properly reset when that node went back to sync mode?

What do you mean `properly reset`? or more exactly, what's `properly`?

@antiochp:
> This checks if we have seen and processed these headers before, not that they are on the header chain. All of these headers could be on a fork and not on the chain itself.

But anyway, now we're in `HeaderSync` state again, and we receive Headers, and all of them **known**, no mater they are on header chain or on a fork. Shouldn't we update `sync_head` at this time?

> I think so yes. We don't want a sync to be running periodically in the background all the time.
We only want to initiate a sync if we know for sure that a peer is advertising at least several blocks ahead of where we believe the head to be.

Okay, Thanks to explain this. Now I agree this:)

> We only want to reset the sync_header when we start a new sync. If we reset it outside of sync then we risk flip-flopping between two or more different forks.

Okay, understand. But I don't find the code to do `reset the sync_header when we start a new sync`, Where is it? and reset to what?

> The sync_head and the header_head are closely linked. They should move is tandem during sync but outside of sync the header_head will update as we process new blocks and the sync_head will get left behind. The sync_head will be stale so needs resetting to the correct header when we transition to sync.

Okay, agree. And same question about where's the code to do this job: `The sync_head will be stale so needs resetting to the correct header when we transition to sync.`?  and then what's the `correct header` at this time?

**Comment by garyyu** [comment]
Sorry, this should be revised:
>  But anyway, now we're in HeaderSync state again, and we receive Headers, and all of them known, no mater they are on header chain or on a fork. Shouldn't we update sync_head at this time?

We need check an additional condition and only update `sync_head` when we're sure it's not a fork. Let me think how to detect it.

**Comment by garyyu** [comment]
I revert the change of `headers_received()` in `adaptor.rs`, and switch to @antiochp proposed solution, which is more simple and efficient.

Please take a new code review, thanks~

File: servers/src/grin/sync.rs [hunk 1]
```diff
@@ -516,16 +519,31 @@ fn get_locator(
...
+	if header_head.height > tip.height {
+		debug!(
+			LOGGER,
+			"sync: get_locator reset to header_head. sync_head: {} at {}, header_head: {} at {}",
+			tip.hash(),
+			tip.height,
+			header_head.hash(),
+			header_head.height,
+		);
+		tip = header_head.clone();
```
[the remark this answers is not in the document]
Comment (antiochp) [reply servers/src/grin/sync.rs]: Let me put a PR up to illustrate what I mean.

[the remark this answers is not in the document]
Comment (garyyu) [reply servers/src/grin/sync.rs]: OK @antiochp  ,wait your PR.
but I'm not sure it's necessary here to update `sync_head`, since we will get the Headers downloaded and update `sync_head` when handle the received Headers.

**Comment by antiochp** [comment]
PR up at #1539. @garyyu

File: servers/src/grin/sync.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (antiochp) [reply servers/src/grin/sync.rs]: The point is we want to do this first.
We're starting a new header sync - we should put the sync_head in the correct state up front.
There's no point sending and receiving 500 headers from another peer, just to then update the stale sync_head to the header_head (that we already know).

**Comment by garyyu** [comment]
The final fix solution in this PR:

	when sync_head is left behind, use header_head as the sync_head for get_locator

The good point is we don't need pollute `sync_head` which is updated ONLY when we receive Headers from peers.

And it's simple, just 3 lines of code (not including the log):
```
fn get_locator(
	...
	if header_head.height > tip.height {
		tip = header_head.clone();
	}
	...
```

But please feel free to use @antiochp 's fix solution  #1539 😃

File: servers/src/grin/sync.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (garyyu) [reply servers/src/grin/sync.rs]: > There's no point sending and receiving 500 headers from another peer, just to then update the stale sync_head to the header_head (that we already know).

No, since we have use `header_head` for `get_locator()`, the new received Headers will all new for us, no duplicated Headers will be sent.

> Or temporarily overriding it when we know it will get updated next iteration.

Sure we know it will get updated next iteration, but this is not a temporarily overriding, it's for the purpose to keep `sync_head` ONLY updated when we received NEW Headers, and we can always count on this `sync_head` to convince us what's the latest Headers we received.

**Comment by antiochp** [comment]
@garyyu This still won't work for long forks though. There are scenarios where `sync_head` and `header_head` must diverge (and remain diverged on different forks) for this to behave properly. 

* Say we are height `1,000` on the main chain.
  * For sake of this example lets just ignore total work and base "most work" chain purely on height.
* Now a peer advertises a previously unknown fork with height `1,100`.
  * So we attempt to sync against this new fork.
* We build a locator and request headers from this peer.
  * We receive ~500 headers from this peer.
* We discover this new fork diverged from the main chain way back at height `400`.
  * The forked headers are `[401', 402', 403', ..., 899', 900']` and we process these.
* At this point - 
  * header_head points to the header at height `1,100`.
  * `sync_head` now points to `900'` (on the forked chain, lower height than `header_head`).

With the logic in this PR we check to see if `sync_head` is "left behind" and we see that it is.
So we ignore it and use `header_head` to build the next locator.
And we end up getting the same headers back in a loop `[401', 402', 403', ..., 899', 900']`.

In this scenario we need to reset `sync_head` _once_ on entering HeaderSync.
And then leave it following the new fork with the aim of it eventually passing our `header_head`, at which point we will switch from following the old chain to the fork which is now the most work chain.

The problem still exists if we use "total work" and not simply "height" when comparing the various forks. We can still be in situations where `sync_head` needs to be temporarily tracking a "less work so far" fork on its way to sync'ing a fork that will eventually have "most work".

I closed #1539 as it was only to demonstrate the intent here. But I don't think we can do it here in the Locator as we do not have enough context to know when to use `sync_head` and when to ignore it, using just the relative heights (or even the relative work as the same problem exists here).

## tl;dr

If we encounter a long fork (in excess of a single 500 header sync msg), `sync_head` needs to track the header sync _on the fork_, while `header_head` will remain on the current chain, until we have enough headers on the fork to make the switch and follow it as the main chain.

**Comment by antiochp** [comment]
> But even we indeed reset the sync_head exactly before starting HeaderSync, this problem will still happen.

On the next iteration `sync_head` will be at `900'` (on the fork) and we will build the next Locator based on this.
And we will request headers from a "most work" peer who will have headers from `900'` to `1100'`.

**Comment by garyyu** [comment]
> On the next iteration sync_head will be at 900' (on the fork)

Yes.

> and we will build the next Locator based on this.

No. Remember that we're talking about `reset the sync_head exactly before starting HeaderSync`. So, next Locator still be the [1000, ...], same as previous iteration.

Anyway, in both your solution and my solution, we can't solve this new problem. `sync_head` is at 900', and `header_head` is at 1000, if we use 900' as tip (the current version), we can't solve the problem in this issue, but your new problem case will not be a problem; if we use 1000 as tip (both your solution and mine solution), we can't solve your new problem case.

So far, the only solution I can see which can solve both problems is what I said at above: (and repeat here)  in case of `sync_head` is different from `header_head`, we need send 2 Headers Request, one use `sync_head` for get_locator, and another one use `header_sync` for get_locator.

**Comment by antiochp** [comment]
I feel like we're going around in circles and that there is some misunderstanding here.

We only transition `NoSync -> HeaderSync` once when we discover we need to start syncing.
Once we are in `HeaderSync` state we stay in this state until we have sync'd the headers completely and the `header_head` has been updated to the head of the most work header chain.

We should only reset `sync_head` on that initial transition to `HeaderSync` state, not on each iteration while we are in this state.

Does that make sense?

**Comment by garyyu** [comment]
> We should only reset `sync_head` on that initial transition to `HeaderSync` state, not on each iteration while we are in this state.

👍👍👍

**Comment by garyyu** [comment]
@antiochp  new code is ready for review.

Now I understand why you insist to reset `sync_head` 😄  So, I just copy your closed PR but give a little bit modification to handle the `initial transition`, and also clearing the history locators.

File: servers/src/grin/sync.rs [hunk 2]
```diff
@@ -141,18 +141,34 @@ pub fn run_sync(
...
-
 							let status = sync_state.status();
-							match status {
-								SyncStatus::TxHashsetDownload => (),
-								_ => {
-									sync_state.update(SyncStatus::HeaderSync {
-										current_height: header_head.height,
-										highest_height: si.highest_height,
-									});
+							let update_sync_state = match status {
```
Comment (ignopeverell) [inline servers/src/grin/sync.rs]: Perhaps move this section to its own function to keep the loop relatively clean?
